### ISSUE TITLE
'my()' incorrectly scoped in 'xxx if yyy' statement.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4642]
Migrated from [rt.perl.org#7982](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7982) (status was 'resolved')

Searchable as RT7982$

**Comment by p5pRT** [comment]
### From stevep@marketview.co.nz

#### Created by stevep@marketview.co.nz



The following code should\, in my opinion\, print five lines\, with the
word 'quux' once on each of them\.

&nbsp;    use strict;
&nbsp;    use warnings;

&nbsp;    sub foo \{

&nbsp;        my $bar = 'baz' if $\_\[0\];
&nbsp;        $bar \.= 'quux';

&nbsp;        return $bar;
&nbsp;    \}

&nbsp;    print foo\(0\)\, "\\n";
&nbsp;    print foo\(0\)\, "\\n";
&nbsp;    print foo\(0\)\, "\\n";
&nbsp;    print foo\(0\)\, "\\n";
&nbsp;    print foo\(0\)\, "\\n";


Instead it prints&#8203;:

&nbsp;    quux
&nbsp;    quuxquux
&nbsp;    quuxquuxquux
&nbsp;    quuxquuxquuxquux
&nbsp;    quuxquuxquuxquuxquux

This behaviour violates DWIM\, and/or 'use strict'\.

It violates DWIM because I wouldn't expect 'my\(\)' to be scoped in this
context \- I expected it to be equivalent to the following snippet\.

&nbsp;    my $bar;
&nbsp;    $bar = 'baz' if $\_\[0\];

Even if it was expected that 'my\(\)' would work like that\, then it
breaks 'use strict'\, since $bar should no longer exist after that
line\, and wouldn't be accessible in the $bar \.= 'quux' line\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.1:

Configured by stevep at Wed Sep  5 14:14:16 NZST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.2.17-14, archname=i686-linux
    uname='linux spinneret 2.2.17-14 #1 mon feb 5 17:53:36 est 2001 i686
unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef
usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing'
    ccversion='', gccversion='egcs-2.91.66 19990314/Linux (egcs-1.1.2
release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
   


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i686-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/stevep
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/stevep/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash2


```
</details>

**Comment by p5pRT** [comment]
### From @tamias

On Wed\, Dec 05\, 2001 at 06&#8203;:34&#8203;:07PM \+1300\, Steve Piner wrote&#8203;:
> 

Thank you for your bug report\.

> The following code should\, in my opinion\, print five lines\, with the
> word 'quux' once on each of them\.
> 
>     use strict;
>     use warnings;
> 
>     sub foo \{
> 
>         my $bar = 'baz' if $\_\[0\];

my has both a compile time effect and a run time effect\.  The compile time
effect declares the variable and allocates storage\.  The run time effect
clears out the storage when the scope is entered\.  By putting a conditional
on the my\, you are skipping the run time effect\, so the previous value
stays around\.

This behavior was unintentional\, but there is now code that depends on it\,
so it probably won't be changed anytime soon\.

In conclusion\, don't do that\.  :\)


Ronald
